backend/telemetry_shard/src/aggregator.rs

```diff
@@ -0,0 +1,271 @@
+
+        // Just as an optimisation, we can keep track of whether we're connected to the backend
+        // or not, and ignore incoming messages while we aren't.
+        let mut connected_to_telemetry_core = false;
+
+        // A list of close channels for the current connections. Send an empty tuple to
+        // these to ask the connections to be closed.
+        let mut close_connections: Vec<mpsc::Sender<()>> = vec![];
+
+        // Maintain mappings from the connection ID and node message ID to the "local ID" which we
+        // broadcast to the telemetry core.
+        let mut to_local_id = AssignId::new();
+
+        // Any messages coming from nodes that have been muted are ignored:
+        let mut muted: HashSet<ShardNodeId> = HashSet::new();
+
+        // Now, loop and receive messages to handle.
+        while let Some(msg) = rx_from_external.next().await {
+            match msg {
+                ToAggregator::ConnectedToTelemetryCore => {
+                    // Take hold of the connection closers and run them all.
+                    let closers = close_connections;
+
+                    for mut closer in closers {
+                        // if this fails, it probably means the connection has died already anyway.
+                        let _ = closer.send(());
+                    }
+
+                    // We've told everything to disconnect. Now, reset our state:
+                    close_connections = vec![];
+                    to_local_id.clear();
+                    muted.clear();
+
+                    connected_to_telemetry_core = true;
+                    log::info!("Connected to telemetry core");
+                }
+                ToAggregator::DisconnectedFromTelemetryCore => {
+                    connected_to_telemetry_core = false;
+                    log::info!("Disconnected from telemetry core");
+                }
+                ToAggregator::FromWebsocket(
+                    _conn_id,
+                    FromWebsocket::Initialize { close_connection },
+                ) => {
+                    // We boot all connections on a reconnect-to-core to force new systemconnected
+                    // messages to be sent. We could boot on muting, but need to be careful not to boot
+                    // connections where we mute one set of messages it sends and not others.
+                    close_connections.push(close_connection);
+                }
+                ToAggregator::FromWebsocket(
+                    conn_id,
+                    FromWebsocket::Add {
+                        message_id,
+                        ip,
+                        node,
+                        genesis_hash,
+                    },
+                ) => {
+                    // Don't bother doing anything else if we're disconnected, since we'll force the
+                    // node to reconnect anyway when the backend does:
+                    if !connected_to_telemetry_core {
```

Comment:
This is probably a terrible idea, but did you consider having a separate loop waiting only for `ConnectedToTelemetryCore` so we can avoid doing even this check?

Reply:
I removed one of these checks (for messages coming form the telemetry core, because clearly it's connected if they come in), so actually there are only 2 now (for add/update nodes). I can't think of a clean way to add a separate bit of handling to pull this out offhand, and it feels like a per-message-type decision to make anyway.

Reply:
Fair.